### PR TITLE
Migrate some tests to buttercup.

### PR DESCRIPTION
**Pull request by cskksc**
Before submitting a PR make sure the following things have been done (and denote this
by checking the relevant checkboxes):

- [ ] The commits are consistent with our [contribution guidelines](CONTRIBUTING.md)
- [ ] You've added tests (if possible) to cover your change(s)
- [x] All tests are passing (`make test`)
- [ ] The new code is not generating bytecode or `M-x checkdoc` warnings
- [ ] You've updated the changelog (if adding/changing user-visible functionality)
- [ ] You've updated the readme (if adding/changing user-visible functionality)
- [ ] You've updated the refcard (if you made changes to the commands listed there)

Thanks!

### CODE REVIEW

File: test/cider-overlay-tests.el [hunk 1]
```diff
@@ -19,47 +19,63 @@
...
+  (describe "overlay duration checks"
+    (describe "when overlays should be erased after next command"
+      (it "next command is executed"
+        (cider--with-overlays
+          (cider--make-result-overlay "ok" :duration 'command)
+          (run-hooks 'post-command-hook)
+          (run-hooks 'post-command-hook)
+          (expect (overlay-position) :to-equal nil)))
+
+      (it "TBD"
```
Comment (cskksc) [inline test/cider-overlay-tests.el]: @Malabarba Why do we have to run the post command hooks twice ? Is it like, once for `make-overlay`, and once for some other command which should hide the overlay ?
Also, what does setting `this-command` to nil do ? Emacs documentation says `"The command now being executed"`, but nothing about nil case. Does it mean `command execution completed` ?

File: test/cider-stacktrace-tests.el [hunk 2]
```diff
@@ -1,33 +1,55 @@
...
-  (let ((cider-stacktrace-suppressed-errors '("a" "b" "g" "j")))
-    (should (cider-stacktrace-suppressed-error-p "a"))
-    (should (cider-stacktrace-suppressed-error-p "b"))
-    (should (cider-stacktrace-suppressed-error-p "g"))
-    (should (cider-stacktrace-suppressed-error-p "j"))
-    (should-not (cider-stacktrace-suppressed-error-p "c"))))
+(describe "cider-stacktrace-some-suppressed-errors-p"
+  :var (cider-stacktrace-suppressed-errors)
+
+  (describe "identifies errors associated with supressed errors"
```
Comment (bbatsov) [inline test/cider-stacktrace-tests.el]: suppressed

File: test/cider-overlay-tests.el [hunk 3]
```diff
@@ -19,47 +19,63 @@
...
+        (cider--with-overlays
+          (setq this-command nil)
+          (cider--make-result-overlay "ok" :duration 'command)
+          (run-hooks 'post-command-hook)
+          (expect (overlay-position) :to-equal nil))))
+
+    (it "when duration is given in secs"
```
Comment (bbatsov) [inline test/cider-overlay-tests.el]: This doesn't sound like a test case description.

File: test/cider-client-tests.el [hunk 4]
```diff
@@ -19,97 +19,128 @@ SYMBOL is locally let-bound to the current buffer."
...
-  (with-connection-buffer "clj" bb1
-    (with-connection-buffer "cljs" bb2
-      ;; The two buffers above are not used. We just want to ensure we are
-      ;; returning the most recent connection in case of ambiguity.
-      (with-connection-buffer "clj" b1
-        (with-connection-buffer "cljs" b2
+(describe "cider-current-connection"
+  :var (cider-connections)
+
+  (it "when there are no active connections"
```
Comment (bbatsov) [inline test/cider-client-tests.el]: We should explain here what's supposed to happen. An `it` shouldn't have a description starting with `when` - this is `describe`'s role. There are some great tips on writing good tests here http://betterspecs.org/

Comment (cskksc) [inline test/cider-client-tests.el]: That is a great link! I was hoping to get the effect of a `context` block by using `when` statements. But yeah, it is incomplete without the supporting `it` description blocks and should go in a `describe` block instead. Will change it today.

On a totally unrelated note, good luck for the talk tomorrow. Have a great one :-) I guess GH PR comment is not the best place for this 😛

Comment (bbatsov) [inline test/cider-client-tests.el]: > On a totally unrelated note, good luck for the talk tomorrow. Have a great one :-) I guess GH PR comment is not the best place for this 😛

Thanks! :-)

Comment (cskksc) [inline test/cider-client-tests.el]: Done.

File: test/cider-overlay-tests.el (hunk 3 above)
Comment (cskksc) [inline test/cider-overlay-tests.el]: Done.

File: test/cider-stacktrace-tests.el (hunk 2 above)
Comment (cskksc) [inline test/cider-stacktrace-tests.el]: Done.

File: test/cider-stacktrace-tests.el [hunk 5]
```diff
@@ -1,33 +1,55 @@
...
-    (should (cider-stacktrace-suppressed-error-p "a"))
-    (should (cider-stacktrace-suppressed-error-p "b"))
-    (should (cider-stacktrace-suppressed-error-p "g"))
-    (should (cider-stacktrace-suppressed-error-p "j"))
-    (should-not (cider-stacktrace-suppressed-error-p "c"))))
+(describe "cider-stacktrace-some-suppressed-errors-p"
+  :var (cider-stacktrace-suppressed-errors)
+
+  (describe "identifies errors associated with suppressed errors"
+    (it "when no errors are suppressed"
```
Comment (cskksc) [inline test/cider-stacktrace-tests.el]: Forgot to look at this file. Changing it now.

File: test/cider-client-tests.el [hunk 6]
```diff
@@ -19,97 +19,146 @@ SYMBOL is locally let-bound to the current buffer."
...
+                (expect (cider-current-connection) :to-equal b1))))
+
+          ;; for cljs
+          (with-connection-buffer "cljs" b1
+            (with-connection-buffer "clj" b2
+              (with-temp-buffer
+                (setq major-mode 'clojurescript-mode)
+                (expect (cider-current-connection) :to-equal b1))))))
+
+      (describe "when a connection matching current file extension doesnt exist"
```
Comment (bbatsov) [inline test/cider-client-tests.el]: doesn't

File: test/cider-client-tests.el [hunk 7]
```diff
@@ -19,97 +19,146 @@ SYMBOL is locally let-bound to the current buffer."
...
-          (should (equal (cider-other-connection b1) b2)))))))
+        (expect (cider-other-connection) :to-equal nil)
+        (expect (cider-other-connection b1) :to-equal nil))
+      ;; for cljs
+      (with-connection-buffer "cljs" b1
+        (expect (cider-other-connection) :to-equal nil)
+        (expect (cider-other-connection b1) :to-equal nil))))
+
+  (describe "when active connections are available"
+    (describe "when a connection of other type doesnt exist"
```
Comment (bbatsov) [inline test/cider-client-tests.el]: doesn't

File: test/cider-util-tests.el [hunk 8]
```diff
@@ -1,31 +1,35 @@
...
+(describe "cider--version"
+  :var (cider-version cider-codename pkg-info-version-info)
 
-(ert-deftest cider--version-snapshot-version ()
-  (noflet ((pkg-info-version-info (package) "0.11.0snapshot (package: 20160301.2217)"))
-    (let ((cider-version "0.11.0-snapshot")
-          (cider-codename "Victory"))
-      (should (string= (cider--version) "0.11.0snapshot (package: 20160301.2217)")))))
+  (it "handles version unavailable error"
+    (fset 'pkg-info-version-info (lambda (package) (error "No version")))
```
Comment (bbatsov) [inline test/cider-util-tests.el]: Won't it be better to just stub the function in question here?

**Comment by bbatsov** [comment]
Overall the PR looks pretty good now. Apart from my small inline remark - please, drop the `.` in the commit message's title.

**Comment by cskksc** [comment]
Done.

**Comment by bbatsov** [comment]
👍

**Pull request merged**